Advance C# linter based on mcs -t:module

hernot opened the pull request:
<!--
READ THIS: Before creating a pull request, please consider the following first.

* The most important thing you can do is write tests. Code without tests
  probably doesn't work, and will almost certainly stop working later on. Pull
  requests without tests probably won't be accepted, although there are some
  exceptions.
* Read the Contributing guide linked above first.
* If you are adding a new linter, remember to update the README.md file and
  doc/ale.txt first.
* If you add or modify a function for converting error lines into loclist items
  that ALE can work with, please add Vader tests for them. Look at existing
  tests in the test/handler directory, etc.
* If you add or modify a function for computing a command line string for
  running a command, please add Vader tests for that.
* Generally try and cover anything with Vader tests, although some things just
  can't be tested with Vader, or at least they can be hard to test. Consider
  breaking up your code so that some parts can be tested, and generally open up
  a discussion about it.
* Have fun!
-->

Added addtional c# linter called mcsc using the module target to not only catch syntax errors but also indicate compilation errors and possibly missing assemblies etc. the lint_file flag is set to 1 to avoid delays on large directory trees. 

Still missing is a possibility to define subdirectories which should be excluded from linting.

And suggestions for improvement wellcome.

Have added a ale-cs.txt file in doc describing the existing mcs and mcsc linter. as this is not included in the files on github.

w0rp commented on ale_linters/cs/mcsc.vim:
```diff
@@ -0,0 +1,86 @@
+    \    . ' "' . join(glob('**/*.cs',v:false,v:true),'" "') . '"'
+	exe 'cd ' . l:cwd
+	return l:cmd
+endfunction
```
This is some very poorly formatted code. Lose all of the type checking, and use blank lines and comments here or there.

w0rp commented on ale_linters/cs/mcsc.vim:
```diff
@@ -0,0 +1,86 @@
+    \    . ' ' . l:assemblies
+	\    . ' -out:' . l:out
+    \    . ' -t:module'
+    \    . ' "' . join(glob('**/*.cs',v:false,v:true),'" "') . '"'
```
Don't use `glob` like this. It's painfully slow on many machines.

w0rp commented on doc/ale-cs.txt:
```diff
@@ -0,0 +1,75 @@
+  the assembly files specified by |g:ale_cs_mcsc_assemblies| or 
+  |b:ale_cs_mcsc_assemblies|. The mcs -unsafe option is set implicitly and has
+  not to be added using |g:ale_cs_mcsc_options| or |b:ale_cs_mcsc_options|
+  variable.
```
The formatting of the text is very strange here. It looks like things are randomly indented. I find it hard to read.

w0rp commented on doc/ale-cs.txt:
```diff
@@ -0,0 +1,75 @@
+  line using the -lib: flag of mcs.
+
+g:ale_cs_mcsc_assemblies                            *g:ale_cs_mcsc_assemblies*
+                                                   *b:ale_cs_mcsc_assemblies*
```
You need to align your help tags to the right margin.

w0rp commented on doc/ale-cs.txt:
```diff
@@ -0,0 +1,75 @@
+  not empty the list of assemblies will be added to the mcsc command
+  line using the -r: flag of mcs. To change the search path mcs uses to 
+  locate the specified assembly files use |g:ale_cs_mcsc_assembly_path| or
+  |b:ale_cs_mcsc_assembly_path| variables
```
Try and explain things more succinctly, and use short sentences.

w0rp commented on ale_linters/cs/mcsc.vim:
```diff
@@ -0,0 +1,86 @@
+\   'executable': 'mcs',
+\   'command_callback': 'ale_linters#cs#mcsc#GetCommand',
+\   'callback': 'ale_linters#cs#mcsc#Handle',
+\	'lint_file': 1
```
Bad indentation.

w0rp commented on README.md:
```diff
@@ -76,7 +76,7 @@ formatting.
 | C++ (filetype cpp) | [clang](http://clang.llvm.org/), [clangcheck](http://clang.llvm.org/docs/ClangCheck.html) !!, [clangtidy](http://clang.llvm.org/extra/clang-tidy/) !!, [cppcheck](http://cppcheck.sourceforge.net), [cpplint](https://github.com/google/styleguide/tree/gh-pages/cpplint) !!, [gcc](https://gcc.gnu.org/), [clang-format](https://clang.llvm.org/docs/ClangFormat.html)|
 | CUDA | [nvcc](http://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html) |
-| C# | [mcs](http://www.mono-project.com/docs/about-mono/languages/csharp/) |
+| C# | [mcs](http://www.mono-project.com/docs/about-mono/languages/csharp/) see:`help ale-cs-mcs` for details,[mcsc](http://www.mono-project.com/docs/about-mono/languages/csharp/) see:`help ale-cs-mcsc` for details and configuration|
```
You need to mark this with `!!`, as you've set `lint_file` to `1`.

You need to document this in `doc/ale.txt` too.

hernot commented:
Ok i have addressed all your recommendations. The only one i at max be able to address partially, if i get some instruction and support form your side is the writing of vader tests. I do not know vader and i do have no docker setup here. So even if able to write a test, i fear I would not be able to verify that it is really testing what it should test. Any advice? Any suggestions?

w0rp commented on ale_linters/cs/mcsc.vim:
```diff
@@ -0,0 +1,107 @@
+
+    " switch back to the working directory to the path
+    " stored above.
+    exe 'cd ' . l:cwd
```
You shouldn't execute a command to change the working directory. Just include `cd ... &&` in the command you want ALE to run.

w0rp commented on ale_linters/cs/mcsc.vim:
```diff
@@ -0,0 +1,107 @@
+    " it seems not to work reliably, which is why vim built in
+    " glob function is used to collect the list of '*.cs'
+    " files.
+    let l:cmd = 'cd ' . l:base . ';'
```
You can just use `return` here instead, with the advice above.

w0rp commented on ale_linters/cs/mcsc.vim:
```diff
@@ -0,0 +1,107 @@
+    \    . 'mcs -unsafe'
+    \    . ' ' . ale#Var(a:buffer, 'cs_mcsc_options')
+    \    . ' ' . l:path
+    \    . ' ' . l:assemblies
```
Try and include these additional arguments only if they aren't empty. Have a look at other linters for examples.

w0rp commented on ale_linters/cs/mcsc.vim:
```diff
@@ -0,0 +1,107 @@
+    \    . ' ' . l:assemblies
+    \    . ' -out:' . l:out
+    \    . ' -t:module'
+    \    . ' "' . join(glob('**/*.cs',v:false,v:true),'" "') . '"'
```
`glob` will perform very badly on many machines. Can you pass just the one file instead, or something like that?

w0rp commented on ale_linters/cs/mcsc.vim:
```diff
@@ -0,0 +1,107 @@
+    let l:cwd = expand('#' . a:buffer . ':p:h')
+    " get the base path of the source tree to lint and if
+    " not empty and exists switch working directory to it
+    let l:base = ale#Var(a:buffer,'cs_mcsc_source')
```
Add a space after commas, here and above.

w0rp commented on ale_linters/cs/mcsc.vim:
```diff
@@ -0,0 +1,107 @@
+    " get the base path of the source tree to lint and if
+    " not empty and exists switch working directory to it
+    let l:base = ale#Var(a:buffer,'cs_mcsc_source')
+    if isdirectory(l:base)
```
Add a blank line above if statements which do not start a new block.

w0rp commented on ale_linters/cs/mcsc.vim:
```diff
@@ -0,0 +1,107 @@
+" list of assemblies to consider
+let g:ale_cs_mcsc_assemblies = get(g:,'ale_cs_mcsc_assemblies',[])
+function! ale_linters#cs#mcsc#GetCommand(buffer) abort
+
```
Delete the blank line here.

w0rp commented on ale_linters/cs/mcsc.vim:
```diff
@@ -0,0 +1,107 @@
+" general mcs options which are likely to stay constant across
+" source trees like -pkg:dotnet
+let g:ale_cs_mcsc_options = get(g:, 'ale_cs_mcsc_options', '')
```
You might like to use `ale#Set` for these variables. Look at the codebase for other examples.

hernot commented:
Hi 
I would appreciate if that workaround would not be necessary but with my installed version of mono mcs compiler the mcs compiler hang nine times out of 10 when trying to use the built in -recurse flag. If it would have worked 10 out of 10 times i would have used that instead of the glob function of vim without resorting to the cd solution if you have better idea for how to properly implement the workaround please let me know.

I decided against including the ale_cs_mcsc_source path inside the glob pattern  to keep command string as short as possible even if several hundred(* files have to be included in the build. 

(*)addmitted a rather poorly organized source tree and and likely indicating of design flaws, still if not all files are included in the linting command (see above) users would receive errors about missing assemblies for classes, types and namespaces defined by source files from the same source tree.

Currently i do have not the time and resources to investigating the reason for mcs hanging when using -recurse flag, in worst submitting a bugreport or even a bugfix to mono project and than adding version checks disabling linter if version with broken -recurse is installed.

hernot commented:
What the hack, now retrying the -recurse flag every thing works as expected. No idea why, what and where the problem was. Need to keep an eye on this. Never the less for the next commit i simplified to using -recurse flag of mcs compiler. Added a NOTE to the help that the case that mcsc does not indicate errors in the currently editor is related to mcs is a multipass comiler which stops after found errors at a specific comilation stage, not advancing to the next. 

Will have a look into the tests, hope i do not make things even worse than now

hernot commented:
a question does a test_mcs_handerl.vader not exist or was it not commited as the mcs linter was already present in the github master branch. If not present at all should i also add a vader test for the mcs like i did for the help files or just for the mcsc i have added?

hernot commented:
Ok added the tests for mcs.vim which was missing them in the master tree and for mcsc.vim
The  continuous-integration/travis-ci/pr test seems from what i can see fail cause i have inserted mcsc in ale-vim.txt file. not sure. please check your self and let me know if that would be the only reason.

hernot commented:
Finally, fixe some nasty behavior, when vim is not run from base directory of source tree.

w0rp commented on test/command_callback/test_cs_mcsc_command_callbacks.vader:
```diff
@@ -0,0 +1,86 @@
+
+  let b:command = ale_linters#cs#mcsc#GetCommand(bufnr(''))
+
+  Assert match( b:command, '^\s*cd\s\+"."\s*;\s*mcs\s\+-unsafe\s\+-out:\S\+\s\+-t:module\s\+-recurse:"\*\.cs"\s*$' ) == 0,
```
The regular expressions here are very hard to read, so it's hard to make sense of what the tests are testing. I recommend writing tests for exact strings with `AssertEqual` instead. You will probably need to build certain parts of the strings base on the filename, etc.

hernot replied:
Is in this case not possible cause the GetCommand registers a temporary file with ale, which differs at every call so the assert is the only possibility to make test succeed, unless in your test machinery you have an AssertFunction or some other function which helps to handle exactly that case.

w0rp replied:
I recommend replacing just the temporary file in the actual string with some other sequence of characters. Then you can use an exact string for the expected string.

w0rp replied:
You could substitute `-out:[^ ]*` or something like that with `-out:TEMP` or just anything really.

hernot replied:
you mean 

let b:command = subsitute(b:command,'/\+tmp/\+\S\+','/tmp/some_tmp_file.tmp')

and than use this modified string in the AssertEqual

w0rp replied:
Yeah, something like that. But bear in mind that the directory will be different from `/tmp` on Mac and Windows, so you should probably search for some characters after `'-out:'`.

hernot replied:
Sure i do. I finish the tests and the fixes commit and as soon as they are working properly i let you know

w0rp commented on ale_linters/cs/mcsc.vim:
```diff
@@ -0,0 +1,85 @@
+\   'executable': 'mcs',
+\   'command_callback': 'ale_linters#cs#mcsc#GetCommand',
+\   'callback': 'ale_linters#cs#mcsc#Handle',
+\    'lint_file': 1
```
The indentation is wrong here.

w0rp commented on ale_linters/cs/mcsc.vim:
```diff
@@ -0,0 +1,85 @@
+    " path and not just the file loaded in the buffer
+    let l:pattern = '^\(.\+\.cs\)(\(\d\+\),\(\d\+\)): \(.\+\): \(.\+\)'
+    let l:output = []
+    let l:source = ale#Var(a:buffer,'cs_mcsc_source')
```
Add a single space after commas.

w0rp commented on ale_linters/cs/mcsc.vim:
```diff
@@ -0,0 +1,85 @@
+    " if list of assemblies to link is not empty convert it to the
+    " appropriate -r: parameter of mcs
+    let l:assemblies = ale#Var(a:buffer,'cs_mcsc_assemblies')
+    if !empty(l:assemblies)
```
Put a single blank line before any if statement which doesn't start a new level of indentation.

w0rp commented on test/command_callback/test_cs_mcsc_command_callbacks.vader:
```diff
@@ -0,0 +1,120 @@
+
+  let b:command = ale_linters#cs#mcsc#GetCommand(bufnr(''))
+  let b:command = substitute(b:command, '-out:' . g:temppathpattern, '-out:' . g:sometempfile, '')
+  let b:command = substitute(b:command, '\s\+', ' ', 'g')
```
Add `unlet! b:command` to your `After` block above. Otherwise this variable might end up being used in some other test.

w0rp commented on test/command_callback/test_cs_mcsc_command_callbacks.vader:
```diff
@@ -0,0 +1,120 @@
+  let g:ale_cs_mcsc_options = '-pkg:dotnet'
+
+  let b:command = ale_linters#cs#mcsc#GetCommand(bufnr(''))
+  let b:command = substitute(b:command, '-out:' . g:temppathpattern, '-out:' . g:sometempfile, '')
```
`sometmpefile` will cause problems here with `\` in paths on Windows. I recommend using something simpler like `'-out:TEMP_FILE` and testing for that. You're replacing it anyway.

hernot commented:
Ok all fixed. 

Also removed Assert match in ale_mcs_handler.vader using a substitute to condense multiple whithespaces into a single space char.

w0rp commented:
Check out the results in Travis CI. `doc/ale.txt` and `README.md` are out of sync.

hernot commented:
The not ins sync problem is not that the tables are not in sync, it is just that the test script stumbles over the greediness of regular expressions in line

 53     | sed 's/see.*\(,\|$\)/\1/g' \

This removes in the following line everything starting at the first see till to the end of the line

| C# | [mcs](http://www.mono-project.com/docs/about-mono/languages/csharp/) see:`help ale-cs-mcs` for details, [mcsc](http://www.mono-project.com/docs/about-mono/languages/csharp/) !! see:`help ale-cs-mcsc` for details and configuration|

with the consequence that only mcs is reported.

I suggest to improve line 53 of the test script as follows

 53     | sed 's/see[^,]*\(,\|$\)/\1/g' \

With this change both tables appear in sync as are. Further down in the README.MD sombody solved the same problem by putting the see of the first entry between (). But is that the desired solution as the the () are already used to indicate the links to the compiler, linter project pages. Further it is not consistent with any other see following the last entry of the lists eg.

| Kotlin | [kotlinc](https://kotlinlang.org) !!, [ktlint](https://ktlint.github.io) !! see `:help ale-integration-kotlin` for configuration instructions

There is no () around the see.  What would be the conformant formatting of the table you suggest markdown would require. In any case i would recommend the fix for properly handling of the greediness of regular expressions.

w0rp commented:
Okay, makes sense. I'll fix it on master.

w0rp commented:
Cheers! :beers:

w0rp commented:
I implemented the regex you suggested. I think that'll do the trick for now.
